### DISCUSSION TITLE
REST API

### DISCUSSION THREAD

**Discussion by ggeop** [ggeop/Python-ai-assistant#111] [Ideas]
This discussion is created for Jarvis REST API development

**Comment by andresh0816** [comment]
@11Tuvork28 At the first, my idea was create an microservice architecture for this Assistant, I mean, we can move the skill recognition to a microservice and the we can separate the rest of functions from the Web API functions.

**Reply by AndromedaCLI** [reply]
Yeah I thought of that too but we would have to most likely rewrite the terminal client. 
I also thought that the core functions of the api to only trigger commands and maybe change settings down the line.

**Reply by andresh0816** [reply]
We need to standarize the skill recognition system. I have one project that it seems to this but made in JS with Python Interoperability. Check it. [Project](https://github.com/leon-ai/leon)

**Reply by AndromedaCLI** [reply]
I will look at it later but I already have something in mind: 
The service would take a query(already as str) execute the skill/or query the wolfram api  and send back the result, I don't really want to implement a system where the audio gets streamed over a websocket to the service but Im open for other ways. 
What do you think?

**Reply by andresh0816** [reply]
Sounds great your abstraction. The audio doesn't need to be streamed over websocket for now, I consider that's kind of things made by the client (Desktop, web or any other.). For now it's ok return the response.

**Reply by AndromedaCLI** [reply]
So we would need to extract the processing of the input to execute an action put an api around it and rework the db implementation. Sounds like a nice weekend :D

**Reply by andresh0816** [reply]
It's easier than it seems. Here I left my discord channel if anyone wants discuss something or just listen music while is coding! [Channel](https://discord.gg/zyH98sBw)

**Reply by AndromedaCLI** [reply]
Yeah, we should break it up into tasks that each of us can work on so that we don't work on the same thing at the same time

**Reply by andresh0816** [reply]
Well, I will take create FastAPI server.

**Reply by AndromedaCLI** [reply]
Sounds good, then I will do the refactoring of the command execution system

**Reply by andresh0816** [reply]
@11Tuvork28 Look at this for your task. Looks like a good alternative for you [Python Library](https://github.com/slapbot/sounder)

**Reply by AndromedaCLI** [reply]
Well I could but I don't intend to change the algorithms at work just the implementation to use them in the api

**Reply by KingofGnome** [reply]
Hey there,
i'm only half a day into this project, still trying to get run properly on Arch Linux.

However, awesome work, and first thing that came into my mind was:
"I'll need Javis on a sever, answering requests/commands from my phone."

Glad to hear you're already working on an API :+1:

**Reply by AndromedaCLI** [reply]
Okay so I came up with the necessary endpoint specifications:
```
@get -> a list with all commands  200
/commands/
{
  id,
  name,
  keywords,
  description,
}
@post
/commands/
{
 keywords,
 arguments,
}
or
{
 raw_user_query,
}
returns 200
{
 command_name,
 answer,
}
If an error occurs then this should be returned:
The appropriate status code
{
 error_code,
 formatted_error_msg,
 raw_error, 
}
raw_error should only be returned when the api is in debugging mode as it is not needed in production mode
```
For handling the execution of commands, there are 2 approaches that come to my mind: 
- The client should only extract keywords/any given arguments and send those to the api.
- The client just sends the raw user query to the api and the api performs all other operations.

Optionally we could add jwt authentication but that would involve the need to create a user(s) and I think that would expand the initial scope way to far.

Let me know what you all think, have a great day :D

**Reply by ggeop** [reply]
Hi @11Tuvork28 ! It looks nice first endpoint design 😁 I suppose that the text to speech functionality will be placed in client side, because as is already mentioned above it's not a best practice to stream voice over a web socket

**Reply by AndromedaCLI** [reply]
Hi @ggeop! Thats correct but theres one problem, how do we distinguish between a client side command like play *some song* and server side command like who is *some person*. Maybe the api response could contain a field with client_side_exec or something like that I don’t know maybe you got an even better idea?

**Reply by KingofGnome** [reply]
Hey folks, just a few thoughts:
First: What about the idea, that the client actually transmits to the server what skills he is capable off. There is no point in responding with "open_libreoffice_ if there is no libreoffice on the client. However, i could also see that "check if possible" completely on the client side.

2nd:
In any way, I think it would make sense to somewhat classify what the skill-return is about. I.e. 
- Skill-Type A is just a text response/information (show on screen or read out loud)
- Skill-Type B is opening a URL
- Skill-Type C is executing a subprocess (prob most of the skills right now?)
- ...

3rd:
To get an API response working, the individual skills must not actually do the action but rather response with the action. A local Jarvis would than have a local handler to fulfill the action, Jarvis on a server would use an api-handler/service to respond the command/text  together with the skilltype so the client knows what to do.

With that kind of an architecture u could have a generic Jarvis-Server and multiple client projects, with different clients for different systems (linux, win, Android,...)

Keep in mind, just a few thoughts from my perspective, i'm still working through the code ;)

**Reply by andresh0816** [reply]
Hey I thought about the client and I'm making a UI design and think create it on Electron.js. For the problem that @11Tuvork28 told us, it's true. We need a way for difference between a server command and client command

**Reply by andresh0816** [reply]
The only idea that I have, it's hardcode the client skills in the respective code, avoiding send an unnecessary request to the server.

**Reply by AndromedaCLI** [reply]
Hi guys long time no hear but I got around to put some work in the project and I got some updates.
Clone my fork and run `docker-compose up -d` and visit http://0.0.0.0:8008/docs to check it out. 
The code is a mess right now but I'm gonna fix that, but still how should we implement client/server commands? I like the idea from @KingofGnome but also hardcoding it would save time :)

**Reply by jameswynn** [reply]
Your fork looks great @11Tuvork28! I was considering doing something very similar, but would also like to break out the TTS and STT pieces to standalone services. Anyway I can help with your refactor more directly?

**Reply by AndromedaCLI** [reply]
Thanks and yeah help would be awesome, if you want you can work with my fork and open a PR once you finished what you're working on. That would be great

**Reply by Red-Blink** [reply]
@11Tuvork28
Did you guys manage to deploy this to the web using flask or whatever? If yes, how can it be done?

**Reply by AndromedaCLI** [reply]
@D3vJohn Nah, currently I focus on other projects that help me learn for my upcoming finals so I don't allocate time to this project sorry

**Comment by StringsTheCatOffical** [comment]
Is there an official tutorial on YouTube for this?
